c2rust-analyze/src/main.rs

```diff
@@ -35,29 +41,113 @@ mod pointer_id;
 mod type_desc;
 mod util;
 
+/// A wrapper around `T` that dynamically tracks whether it's initialized or not.  `RefCell`
+/// dynamically tracks borrowing and panics if the rules are violated at run time; `MaybeUnset`
+/// dynamically tracks initialization and similarly panics if the value is accessed while unset.
+#[derive(Clone, Copy, Debug)]
+struct MaybeUnset<T>(Option<T>);
+
+impl<T> Default for MaybeUnset<T> {
+    fn default() -> MaybeUnset<T> {
+        MaybeUnset(None)
+    }
+}
+
+impl<T> MaybeUnset<T> {
+    pub fn set(&mut self, x: T) {
+        if self.0.is_some() {
+            panic!("value is already set");
+        }
+        self.0 = Some(x);
```

Comment:
```suggestion
        if self.0.replace(x).is_some() {
            panic!("value is already set");
        }
```

Reply:
IMO this makes the logic harder to follow

Reply:
I disagree.  The `Option` methods are designed for exactly this and are idiomatic.

Reply:
fwiw I agree with Stuart that this is harder to follow -- for me it's because I'm less familiar with `replace`

Reply:
There's also `std::mem::{replace,take}`.  It's an idiom Rust uses in multiple places, so I do think it's still worth it to use.  You could use `match` instead of `map` or `expect`, too, as the logic is easier for a beginner to follow, but you would probably encourage people to use the dedicated method and have them learn it.

Reply:
The connotation of `replace` to me is that there are two interesting values, a new one and (possibly) an old one.  Here, we don't care about the old value at all (and in fact we're asserting that there isn't one), so `replace` seems like an unusual choice.

Similarly, `take` is about extracting the value out of an option, so using it when we don't care about the value is a bit weird.  I would guess that this `x.take().unwrap();` type of construct (where the result is unused) is almost unheard of in Rust code.

Overall, I think the original versions of these functions express the high-level intent about as clearly and directly as they possibly could.  Switching to `replace`/`take` brings in concepts of old/new values which are not actually helpful for understanding.

Reply:
[`Cell::set` is implemented](https://doc.rust-lang.org/src/core/cell.rs.html#362) using `Cell::replace`.  That's exactly what we're doing here, using replace to set a new value.  Except unlike `Cell`, we don't care about propagating the old value to the user, but we want to make sure we didn't actually replace a value, so we panic instead.